enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,232 @@
+
+```go
+type cloudProviderObserver struct {
+	targetNamespaceName     string
+	cloudProviderNamePath   []string
+	cloudProviderConfigPath []string
+	observeExternal         bool
+}
+```
+
+The `observeExternal` field will determine, if `status.platformStatus.external` field from the `Infrastructure` resource will be taken in account in establishing needed flags for a Kubernetes binary, such as some `cloud-controller-manager`, `kube-controller-manager`, etc. It will make the observer type to ignore content of the `cloud-config` file, and force the provider to be set as `external` if the option is not selected.
+
+#### Kubelet
+
+Openshift manages `kubelet` configuration with `machine-config-operator`. The actual `cloud-config` arguments are passed externally from the `Infrastructure` resource on https://github.com/openshift/machine-config-operator/blob/2a51e49b0835bbd3ac60baa685299e86777b064f/pkg/controller/template/render.go#L310-L357, where the support for `external` cloud provider will be added. For each platform with `external` configuration, the `cloud-config` flag will not be set.
+
+#### Kube-controller-manager
+
+This configuration is managed by the `cluster-kube-controller-manager-operator` code and will need to integrate the `library-go` changes, described above in the operator [configuration](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/990ea3f2ace13be6579c00a889842c3f5d3e756a/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L82-L85).
+
+#### Kube-api-server
+
+No changes for the current configuration are needed
+
+### Operator resource management strategy
+
+Every cloud manages it's configuration differently, but share a common overall structure.
+
+1. `cloud-controller-manager` and related resources are deployed under one namespace.
+2. Each pod (`ccm`, `cloud-node-manager`, etc.) supports deployment with a `DaemonSet`.
+3. The pods are expected to run on `Nodes` marked as a control-plane with annotation:
+```yaml
+    node-role.kubernetes.io/master: ""
+```
+4. The pods are sharing common command line arguments with the `kcm` pods, such as `--cloud-provider` and `--cloud-config`.
+
+Proposed operator will not diverge from the upstream resource structuring:
+
+- Static infrastructure resources creation, such as namespaces or service accounts will be delegated to the [CVO](https://github.com/openshift/cluster-version-operator).
+- Operator will populate, create and manage the cloud-specific set of `DaemonSets`, based on the selected `platform` in the `Infrastructure` resource.
```

Comment:
There are cloud specific workloads? This is a strong argument against merging it into kcm pod.

Reply:
Yes there are. Aside from `ccm` and `cnm` for Azure, there is also a VSphere with an additional node-port service for connecting to the API endpoint: https://github.com/kubernetes/cloud-provider-vsphere/blob/master/manifests/controller-manager/vsphere-cloud-controller-manager-ds.yaml#L63

Reply:
I agree that this shouldn't be merged into the KCM pod, we should have our own, and have it cloud specific (eg called `aws-cloud-controller-manager` on AWS) so that if we ever decided to run more than one cloud in a cluster, the migration to that approach would not cause conflicts with naming